src/inmanta/execute/runtime.py

```diff
@@ -508,10 +527,15 @@ class BaseListVariable(DelayedResultVariable[ListValue]):
 
     value: "List[Instance]"
 
-    __slots__ = ()
+    __slots__ = ("_listeners", "_done_listeners", "_nb_pure_listeners")
 
     def __init__(self, queue: "QueueScheduler") -> None:
-        self.listeners: List[ResultCollector[ListValue]] = []
+        # keep track of listeners and whether they may cause progress when this variable is frozen
+        # use dict for easy lookup with reliable ordering
+        self._listeners: Optional[dict[ResultCollector[ListValue], None]] = {}
```

Comment:
I expect this to be more efficient because of the cleanup. I'll try to do some compiler scaling tests with this and your other performance related comment in mind.

Reply:
It requires about 1% more memory in the modified compilerscaling test, and the cleanup doesn't seem to have much of an effect so if I dropped it and made this a plain list again.